proposals/compatibility-mode.md

```diff
@@ -0,0 +1,71 @@
+# WebGPU Compatibility Mode
+
+This proposal is **under active development, but has not been standardized for inclusion in the WebGPU specification**. WebGPU implementations **must not** expose this functionality; doing so is a spec violation. Note however, an implementation might provide an option (e.g. command line flag) to enable a draft implementation, for developers who want to test this proposal.
+
+The changes merged into this document are those for which the GPU for the Web Community Group has achieved **tentative** consensus prior to official standardization of the whole propsal. New items will be added to this doc as tentative consensus on further issues is achieved.
+
+## Problem
+
+WebGPU is a good match for modern explicit graphics APIs such as Vulkan, Metal and D3D12. However, there are a large number of devices which do not yet support those APIs. In particular, on Chrome on Windows, 31% of Chrome users do not have D3D11.1 or higher. On Android, [23% of Android users do not have Vulkan 1.1 (15% do not have Vulkan at all)](https://developer.android.com/about/dashboards). On ChromeOS, Vulkan penetration is still quite low, while OpenGL ES 3.1 is ubiquitous.
+
+## Goals
+
+The primary goal of WebGPU Compatibility mode is to increase the reach of WebGPU by providing an opt-in, slightly restricted subset of WebGPU which will run on older APIs such as D3D11 and OpenGL ES. This will increase adoption of WebGPU applications via a wider userbase.
+
+Since WebGPU Compatibility mode is a subset of WebGPU, all valid Compatibility mode applications are also valid WebGPU applications. Consequently, Compatibility mode applications will also run on user agents which do not support Compatibility mode. Such user agents will simply ignore the option requesting a Compatibility mode Adapter and return a Core WebGPU Adapter instead.
+
+## WebGPU Spec Changes
+
+```webidl
+partial dictionary GPURequestAdapterOptions {
+    boolean compatibilityMode = false;
+}
+```
+
+When calling `GPU.RequestAdapter()`, passing `compatibilityMode = true` in the `GPURequestAdapterOptions` will indicate to the User Agent to select the Compatibility subset of WebGPU. Any Devices created from the resulting Adapter on supporting UAs will support only Compatibility mode. Calls to APIs unsupported by Compatibility mode will result in validation errors.
```

Comment:
I think the first part is covered in the "Goals" section above: "Since WebGPU Compatibility mode is a subset of WebGPU, all valid Compatibility mode applications are also valid WebGPU applications. Consequently, Compatibility mode applications will also run on user agents which do not support Compatibility mode."

I've added a sentence about reducing the restrictions to a minimum in the same section.

Reply:
To clarify: do all implementations have to add the compat validation even if a non-compat adapter has been returned?

> Calls to APIs unsupported by Compatibility mode will result in validation errors.

I'm interpreting this as a "yes".

I think this is fine either way but it's worth clarifying since I don't remember this being discussed.

Reply:
Depends what you mean by "non-compat adapter". If the developer requests compatibilityMode, then a supporting UA will return an Adapter that will validate all calls made on it against the Compatibility subset, even if the underlying physical device happens to be D3D12 or Vulkan, for example. Is that what you're asking?

(Nit: to me, such an Adapter is a "compat adapter", not a "non-compat adapter", since it appears to the developer to obey all the same constraints as one backed by OpenGL ES.)

Reply:
So, when a user asks for `compatibilityMode = true` they can either get a "compat adapter" or a "non-compat adapter" depending on whether or not the UA supports the WebGPU Compat mode.

The problem I see is that there is no way for the user to know whether the UA supports the WebGPU Compat mode or not.
I see this being an issue for developers trying to test their compat apps on an UA that doesn't support compat and subsequently not getting any compat validation errors.

Reply:
Was this written this way to make it easy to unship the compat mode in the future? If so, I guess this is fine but we should add a note somewhere in the spec that clarifies the fact that compat validation errors are not guaranteed on UAs that don't support the compat mode.

Reply:
> I see this being an issue for developers trying to test their compat apps on an UA that doesn't support compat and subsequently not getting any compat validation errors.

It's not that difficult to provide a JavaScript extension or JS library that enforces most compat mode validation. I've written an incomplete one [here](https://github.com/greggman/webgpu-dev-extension/blob/main/extension/scripts/webgpu-compat-validation.js) and would be happy to complete it. Not sure if that helps this issue

Reply:
Actually, I don't believe the initial motivation was related to unshipping. We were unsure whether all browser vendors would be willing to make Compatibility mode a required part of the WebGPU spec, or leave it optional. You're right that developers would be unable to test against compatibility mode if their browser did not support it. If all vendors are comfortable making it mandatory, we could reconsider the language above.

That said, I believe it is possible to check if a browser's support by checking the "isCompatibilityMode" property on the returned Adapter.